base/compiler/tfuncs.jl

```diff
@@ -1594,6 +1602,12 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
     end
     if istuple
         return Type{<:appl}
+    elseif isa(appl, DataType) && appl.name === _NAMEDTUPLE_NAME && length(appl.parameters) == 2 &&
+           (appl.parameters[1] === () || appl.parameters[2] === Tuple{})
+        # if the first/second parameter of `NamedTuple` is known to be empty,
+        # the second/first argument should also be empty tuple type,
+        # so refine it here
+        return Const(NamedTuple{(),Tuple{}})
```

Comment:
This reasoning seems faulty, since couldn't the parameter also be a TypeVar of any sort?

Reply:
You're quite right:
```
julia> (()->NamedTuple{(), <:Any})()
NamedTuple{(), Tuple{}}
```